i18n-helpers/src/bin/mdbook-xgettext.rs

```diff
@@ -86,21 +103,161 @@ fn create_catalog(ctx: &RenderContext) -> anyhow::Result<Catalog> {
+            // Iterating through the book in section-order, the PartTitle
+            // represents the 'section' that each chapter exists within.
+            current_top_level = slug(title);
+        } else if let BookItem::Chapter(chapter) = item {
+            let source = match &chapter.path {
                 Some(path) => ctx.config.book.src.join(path),
                 None => continue,
             };
+
+            let template_destination = if depth == 0 {
+                PathBuf::from("messages")
+            } else if depth == 1 {
+                PathBuf::from(current_top_level.clone())
+            } else {
+                // The current chapter is already at depth 2, so append the chapter's
+                // name for depths greater than 1.
+                PathBuf::from(current_top_level.clone()).join(slug(&chapter.name))
+            };
+
+            // Add the (destination, catalog) to the map if it doesn't yet
+            // exist, so messages can be appended to the catalog.
+            if let None = catalogs.get(&template_destination) {
+                catalogs.insert(
+                    template_destination.clone(),
+                    Catalog::new(generate_catalog_metadata(&ctx)),
+                );
+            }
+            let catalog: &mut Catalog = catalogs.get_mut(&template_destination).unwrap();
+
             for (lineno, msgid) in extract_messages(&chapter.content) {
-                let source = format!("{}:{}", path.display(), lineno);
-                add_message(&mut catalog, &msgid, &source);
+                let source_with_lineno = format!("{}:{}", source.display(), lineno);
+                add_message(catalog, &msgid, &source_with_lineno);
+            }
+
+            // Add the contents for all of the sub-chapters within the current chapter.
+            for ChapterTemplateInfo {
+                content,
+                source,
+                template_destination,
+            } in get_subcontent_for_chapter(chapter, template_destination, depth, 2)
+            {
+                if let None = catalogs.get(&template_destination) {
+                    catalogs.insert(
+                        template_destination.clone(),
+                        Catalog::new(generate_catalog_metadata(&ctx)),
+                    );
+                }
+                let catalog = catalogs.get_mut(&template_destination).unwrap();
+                let source = ctx.config.book.src.join(&source);
+
+                for (lineno, msgid) in extract_messages(&content) {
+                    let source_with_lineno = format!("{}:{}", source.display(), lineno);
+                    add_message(catalog, &msgid, &source_with_lineno);
+                }
             }
         }
     }
 
-    Ok(catalog)
+    Ok(catalogs)
+}
+
+/// A view into the relevant template information held by `mdbook::book::Chapter`
+/// and a location to store the exported polib messages.
+struct ChapterTemplateInfo {
```

Comment:
The word "template" here and elsewhere confuses me — possibly because I've done a lot of web development. For me, a "template" is something you later instantiate and fill out with concrete values.

So a "chapter template" is not clear to me since the chapters are already full of content?

Could we call the struct `Chapter` instead since I think it's a minimal version of `BookItem::Chapter` enum variant? The `template_destination` field could then conveniently become `destination`.

Reply:
No problem at all! I can definitely see how template would mean something completely different in that context. `Chapter` is more clear here